### DISCUSSION TITLE
[PHP] Bad Request on /catalog/2022-04-01/items

### DISCUSSION THREAD

**Discussion by danielerainone** [amzn/selling-partner-api-models#3541] [General]
Hi,
I'm trying to integrate my web app with Amazon's SP-API but I'm getting an error that I can't resolve as the response is provided to me in text/html with no error messages or anything else.

This is my provider, my signature, my credentials and the various headers I apply:
```
$this->_provider = new \GuzzleHttp\Client();
$this->_signature = new \Aws\Signature\SignatureV4( 'execute-api', 'us-east-1');
$this->_credentials = new \Aws\Credentials\Credentials( 'LWA_CLIENT', 'LWA_SECRET' );

$this->_headers = [
	'host'   => 'https://sellingpartnerapi-na.amazon.com',
	'x-amz-access-token' => $authProvider->getToken(),
	'x-amz-date' => date( 'Ymd' ),
];
```

This is my my function to elaborate and send request:
```
public function request( string $path, array $Arrayquery = [], array $body = null ) {
		$query = http_build_query( $Arrayquery, '', '&' );

		$request = $this->_signature->signRequest(
			new Request( 'GET', new \GuzzleHttp\Psr7\Uri( "{$this->_host}{$path}?{$query}" ), $this->_headers, $body ),
			$this->_credentials
		);

		$response = $this->_provider->sendRequest( $request );

		if ( $response->getStatusCode() >= 300 ) {
			print_r( $response->getBody()->__toString() );
			//throw new Exception( $response->getReasonPhrase(), $response->getStatusCode() );
		} else {
			return $response->getBody();
		}
	}
```

And this is how i make a request:
```
$provider->request( '/catalog/2022-04-01/items', [
	'marketplaceIds'  => 'APJ6JRA9NG5V4',
	'brandNames'     => 'Samsung',
] );
```

The response I get is a "400 Bad Request" in html format with no error message or anything.

I don't know if I did something wrong in configuring the request, authentication or anything else.

Thank you in advance,
Good day.

**Comment by tomaszhanc** [comment]
For PHP, I'm using this SDK https://github.com/amazon-php/sp-api-sdk, and it works seamlessly.

**Reply by danielerainone** [reply]
HI,
I had also found it but since it was not official and with few stars, I had excluded it and wanted to develop it independently for safety reasons.

Thanks anyway for the feedback, if no one can help me I will certainly use it!

**Reply by stefnats** [reply]
There is also https://github.com/jlevers/selling-partner-api which has quite a few stars.

I suggest you use one of those libraries that are used by others in production instead of re-inventing the wheel and spend too much time on this.

In case you have security concerns, you can decide to look into the source code or in fact implement it yourself, but this comes with a price.

**Reply by tomaszhanc** [reply]
@danielerainone it's not official, but it's auto-generated from official Amazon Open Api models. I would say it's safe to use.

**Reply by danielerainone** [reply]
> There is also https://github.com/jlevers/selling-partner-api which has quite a few stars.
> 
> I suggest you use one of those libraries that are used by others in production instead of re-inventing the wheel and spend too much time on this.
> 
> In case you have security concerns, you can decide to look into the source code or in fact implement it yourself, but this comes with a price.

Thank you both for the advice.
I'm relying on https://github.com/jlevers/selling-partner-api but I'm having trouble creating the products. The ProductTypeDefinitionsV20200901Api returns me a json with all the structure and documentation of a product, but in the "required" object there aren't actually all the mandatory fields, but those connected to the product type are missing.

Does anyone have any idea how I can fetch the required fields based on the Product Type?

A thousand thanks,
Good day.
